### DISCUSSION TITLE
Using getInitialProps within nested layouts that stay mounted between page transitions

### DISCUSSION THREAD

**Discussion by Xetera** [vercel/next.js#14553] [Help] [answered]
Hi I'm currently migrating a react app to nextjs and I'm stuck at a portion with nested routing on a page with tab navigation. I've followed through the dynamic navigation examples and have a `pages` folder thats roughly laid out like
- `/pages/group/[id]/[name]`
- `/pages/group/[id]/[name]/foo`
- `/pages/group/[id]/[name]/bar`

Transitioning between these pages should persist some kind of layout that I can't simply put inside `_app.js` because it should only be persisted on the `/group` path but I can't seem to be able to get that effect without unmounting and re-mounting the entire `/group/[id]/[name]/*` component tree within page transitions between `foo` and `bar` which is both inefficient and resets scroll positions.

I've stumbled across [a great article that explains how to implement this](https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/) while looking things up but I realized that this concept doesn't work if your layouts have some sort of state or data-fetching involved using `getInitialProps`. I've dug around trying to implement this in a library I was using https://github.com/adamsoffer/next-apollo/issues/74 before I realized the problem isn't related to apollo but nextjs in general.

Am I approaching this the wrong way? My goal is to be able to have a shared component between sub-pages that supports `getInitialProps` and doesn't get unmounted between page transitions. I'm kind of confused since the thing I'm trying to do isn't niche and something many react projects do with react-router so I feel like I must be missing some obvious documentation or something else that lets me do this.

**Comment by bduffany** [comment]
I think it might still be possible to combine `getInitialProps` with the persistent layout pattern. Have you tried something like this?

```typescript

// _app.ts

type NextPageWithLayout = NextPage & {
  LayoutComponent: React.FC;
}

export default class CustomApp extends App {
  render() {
    const { Component: Page, pageProps } = this.props;
    const LayoutComponent = (Page as NextPageWithLayout).LayoutComponent || ExampleDefaultLayout;

    return (
      <LayoutComponent>
        <Page {...pageProps}></Page>
      </LayoutComponent>
    );
  }
}
```

```typescript
// pages/group/<foo>.tsx

const GroupPage = ({ exampleServerSideProp }) => {
  return <>{exampleServerSideProp}</>;
}

const GroupPageWithApollo = withApollo(GroupPage);

(GroupPageWithApollo as NextPageWithLayout).LayoutComponent = GroupLayout;

GroupPageWithApollo.getInitialProps = async () => {
  return { exampleServerSideProp: await makeApiRequestFor('/foo/bar'); };
};

export default GroupPageWithApollo;
```

**Reply by bduffany** [reply]
Hmm actually I see you have the `getDataFromTree` call in your post about `next-apollo`. I'm guessing that's what is complicating things for you. What does that do?

**Reply by bduffany** [reply]
I edited my answer above -- what if you put the `Layout` prop on the component that has been wrapped with `withApollo`? Does that work? (I'm interested in this as well because I've been using Apollo without `withApollo` and it has been kinda messy)

**Reply by Xetera** [reply]
The problem with this workaround is that because the `LayoutComponent` isn't a part of the component tree being exported from the page, it never actually gets server side rendered because it doesn't end up inside of `getDataFromTree`. You can't just wrap the exported component around a layout either because nextjs exposes the entire `<AppTree />` where the exported component from the page is wrapped around the entire nextjs tree.

This way the layout is accessible from `_app.tsx` but it means I can't have shared data between pages. I suppose one workaround could be to re-fetch the same data on every sub-page and rely on the apollo to cache the original request which is what I'm probably going to end up doing at this point since this is proving to be really difficul... But of course that adds more complexity, I can't simply move up the data fetching into the pages because the layout passes the fetched data into multiple different child components. I'm really surprised this isn't something more people are having problems with

**Reply by bduffany** [reply]
Is this bug related? https://github.com/vercel/next.js/issues/9336 The OP mentions that the with-apollo examples aren't working because of `<AppTree />` being incorrect

**Comment by TheWuster935** [comment]
Any updates? I also followed the same article about persistent layouts and am running into this issue.

**Comment by Xetera** [comment, marked as answer]
Seeing as the nextjs team isn't interested in prioritizing this, I decided to address this issue by moving to https://github.com/remix-run/remix (which has built in support for nested layouts) after coming across the exact same problem in every single nextjs app I've built in the past 2 years since asking this question. The move ended up also solving every other kind of data fetching problem I've had in nextjs so far which has been a major issue for me in general. Couldn't recommend it more
